src/middleware/builtin.spec.ts

```diff
@@ -0,0 +1,155 @@
+
+describe('matchMessage()', () => {
+  function initializeTestCase(pattern: string | RegExp): Mocha.AsyncFunc {
+    return async () => {
+      // Arrange
+      const { matchMessage } = await importBuiltin();
+
+      // Act
+      const middleware = matchMessage(pattern);
+
+      // Assert
+      assert.isOk(middleware);
+    };
+  }
+
+  function matchesPatternTestCase(pattern: string | RegExp, matchingText: string): Mocha.AsyncFunc {
+    return async () => {
+      // Arrange
+      const dummyContext: DummyContext = {};
+      const { fn: next, promise: onNextFirstCall } = wrapToResolveOnFirstCall(assertions);
+      const fakeArgs = {
+        next,
+        message: createFakeMessageEvent(matchingText),
+        context: dummyContext,
+      } as unknown as MessageMiddlewareArgs;
+      const { matchMessage } = await importBuiltin();
+
+      // Act
+      const middleware = matchMessage(pattern);
+      middleware(fakeArgs);
+      await delay();
+
+      // Assert
+      async function assertions(...args: any[]): Promise<void> {
+        // Assert that there is no error
+        assert.notExists(args[0]);
+        // The following assertion(s) check behavior that is only targeted at RegExp patterns
+        if (typeof pattern !== 'string') {
+          if (dummyContext.matches !== undefined) {
+            assert.lengthOf(dummyContext.matches, 1);
+          } else {
+            assert.fail();
+          }
+        }
+      }
+      return onNextFirstCall;
+    };
+  }
+
+  function notMatchesPatternTestCase(pattern: string | RegExp, nonMatchingText: string): Mocha.AsyncFunc {
+    return async () => {
+      // Arrange
+      const dummyContext = {};
+      const fakeNext = sinon.fake();
+      const fakeArgs = {
+        message: createFakeMessageEvent(nonMatchingText),
+        context: dummyContext,
+        next: fakeNext,
+      } as unknown as MessageMiddlewareArgs;
+      const { matchMessage } = await importBuiltin();
+
+      // Act
+      const middleware = matchMessage(pattern);
+      middleware(fakeArgs);
+      await delay();
+
+      // Assert
+      assert(fakeNext.notCalled);
+      assert.notProperty(dummyContext, 'matches');
+    };
+  }
+
+  function noTextMessageTestCase(pattern: string | RegExp): Mocha.AsyncFunc {
+    return async () => {
+      // Arrange
+      const dummyContext = {};
+      const fakeNext = sinon.fake();
+      const fakeArgs = {
+        message: createFakeMessageEvent([{ type: 'divider' }]),
+        context: dummyContext,
+        next: fakeNext,
+      } as unknown as MessageMiddlewareArgs;
+      const { matchMessage } = await importBuiltin();
+
+      // Act
+      const middleware = matchMessage(pattern);
+      middleware(fakeArgs);
+      await delay();
+
+      // Assert
+      assert(fakeNext.notCalled);
+      assert.notProperty(dummyContext, 'matches');
+    };
+  }
+
+  describe('using a string pattern', () => {
```

Comment:
I'm trying something weird and different here. Looking for feedback. What it boils down to is that instead of writing each of these test cases twice, factoring out the actual test cases into functions that return them given a certain set of inputs (in this case the `pattern`). Does this make sense? It significantly reduces repetition, but not 100%. I chose not to continue factoring the function given to `describe()` into its own test case generator function because I think that might be one step too much of abstraction (and it would look weird because calls to `it()` depend on having the right `this`, so binding would be necessary).

Reply:
I like this test pattern and would advocate for it. It makes it easy to read the different cases for a single function or piece of functionality. 👍